src/YarpUtilities/src/VectorsCollectionServer.cpp

```diff
@@ -0,0 +1,181 @@
+    if (!m_pimpl->port.open(portName))
+    {
+        log()->error("{} Unable to open the port named {}.", logPrefix, portName);
+        return false;
+    }
+
+    // open the rpc port
+    const std::string rpcPortName = remote + "/rpc:i";
+    if (!m_pimpl->rpcPort.open(rpcPortName))
+    {
+        log()->error("{} Unable to open the rpc port named {}.", logPrefix, rpcPortName);
+        return false;
+    }
+
+    if (!VectorsCollectionMetadataService::yarp().attachAsServer(m_pimpl->rpcPort))
+    {
+        log()->error("{} Unable to attach the metadata port.", logPrefix);
+        return false;
+    }
+
+    return true;
+}
+
+bool VectorsCollectionServer::populateMetadata(const std::string& key,
+                                               const std::vector<std::string>& metadata)
+{
+    // check if the metadata has been already populated
+    if (m_pimpl->isMetadataFinalized)
+    {
+        log()->error("[VectorsCollectionServer::populateMetadata] The metadata has been already "
+                     "populated.");
+        return false;
+    }
+
+    // check if the key has been already used
+    if (m_pimpl->setOfKeys.find(key) != m_pimpl->setOfKeys.end())
+    {
+        log()->error("[VectorsCollectionServer::populateMetadata] The key {} has been already "
+                     "used.",
+                     key);
+        return false;
+    }
+
+    // populate the metadata
+    m_pimpl->metadata.vectors[key] = metadata;
+
+    // add the key to the set of keys
+    m_pimpl->setOfKeys.insert(key);
+
+    return true;
+}
+
+bool VectorsCollectionServer::finalizeMetadata()
+{
+    // check if the metadata has been already finalized
+    if (m_pimpl->isMetadataFinalized)
+    {
+        log()->error("[VectorsCollectionServer::finalizeMetadata] The metadata has been already "
+                     "finalized.");
+        return false;
+    }
+
+    // check if the metadata is empty
+    if (m_pimpl->metadata.vectors.empty())
+    {
+        log()->error("[VectorsCollectionServer::finalizeMetadata] The metadata is empty.");
+        return false;
+    }
+
+    // set the metadata as finalized
+    m_pimpl->isMetadataFinalized = true;
+
+    return true;
+}
+
+bool VectorsCollectionServer::populateData(const std::string& key,
+                                           const iDynTree::Span<const double>& data)
+{
+    constexpr auto logPrefix = "[VectorsCollectionServer::setData]";
+
+    // check if the metadata has been finalized
+    if (!m_pimpl->isMetadataFinalized)
+    {
+        log()->error("{} The metadata has not been finalized.", logPrefix);
+        return false;
+    }
+
+    // check if the key exists
+    if (m_pimpl->setOfKeys.find(key) == m_pimpl->setOfKeys.end())
+    {
+        log()->error("{} The key {} does not exist.", logPrefix, key);
+        return false;
+    }
+
+    // prepare the data
+    BipedalLocomotion::YarpUtilities::VectorsCollection& collection = m_pimpl->port.prepare();
```

Comment:
I just realized this. Wasn't it a problem to call ``prepare`` multiple times?

Reply:
Well, I tested and everything seems working fine

Reply:
I dug a little bit inside the ``prepare`` mechanism, and I guess this might explode in memory. Here is the trace of calls:
1. https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/BufferedPort-inl.h#L113
2. https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/PortWriterBuffer.h#L98-L102
3. https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/PortWriterBufferBase.cpp#L185
4. https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/PortWriterBufferBase.cpp#L69-L93
5. https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/impl/PortCorePackets.cpp#L35-L54

In particular, I suspect that the problem is with https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/impl/PortCorePackets.cpp#L37-L42, where, if it does not find any ``inactive`` packet, it allocates a new one.

At this point, to be on the safe side, I would suggest keeping an internal buffer, fill it in with every ``populateData``, and then do the ``prepare`` and a copy in the ``sendData``. What do you think?

Reply:
Actually, it should not explode. I just saw https://github.com/robotology/yarp/blob/5b7014770df1da99c48b4dd27871b5a7a3dc5655/src/libYARP_os/src/yarp/os/PortWriterBufferBase.cpp#L71-L77. In particular 
```
            // oops, there is already a prepared and unwritten
            // object.  best remove it.
```
so I suppose that the multiple ``prepare`` case is already handled.

So feel free to proceed as you prefer